### PR TITLE
Fix bug 1008 : regression in move to numeric edited items

### PR DESCRIPTION
**Pull request by ddeclerck**
This patch by @chuck-haatvedt fixes bug [1008](https://sourceforge.net/p/gnucobol/bugs/1008).

### CODE REVIEW

**Comment by GitMensch** [comment]
Looks good to me. Can you check that in in behalf of chuck? I'll adjust the metadata on Monday. 

But please say in the changelog also a bit what this bug was about.

**Comment by GitMensch** [comment]
See upstream bug report for updated tests and patch

**Comment by ddeclerck** [comment]
> See upstream bug report for updated tests and patch

I see there have been a few more fixes. I'll try to have a look at this before monday.

**Comment by ddeclerck** [comment]
@GitMensch I've included the updates from Chuck ; is it good for merge ?

**Comment by ddeclerck** [comment]
Note: I've also just removed the extra error message:
```
cob_runtime_error ("optimized_move_display_to_edited: overflow in destination field");
```

This message was not present before, and truncation occured silently (this is another "regression" reported by our customer).

**Comment by GitMensch** [comment]
Please add an example to the testsuite where the removed message previously was returned.

From the first look this should indeed not happen, should it? If it is reasonable, then the `break` should possibly be possibly be changed to a `goto endofouterloop`.

As you adjust this function in any case - please move `n = p->times_repeated;` into the `for` loop and, if possible, localize the `int n` into the loop.

**Comment by ddeclerck** [comment]
> Please add an example to the testsuite where the removed message previously was returned.

Done.

> From the first look this should indeed not happen, should it?

This happens when moving to a shorter edited item.

> If it is reasonable, then the `break` should possibly be possibly be changed to a `goto endofouterloop`.

Done.
 
> As you adjust this function in any case - please move `n = p->times_repeated;` into the `for` loop and, if possible, localize the `int n` into the loop.

Also done ;)

File: libcob/move.c [hunk 1]
```diff
@@ -1047,8 +1045,8 @@ optimized_move_display_to_edited (cob_field *f1, cob_field *f2)
 	c = 0x00;
 
 	for (p = COB_FIELD_PIC (f2); p->symbol; ++p) {
+		int n = p->times_repeated;
 		c = p->symbol;
```
Comment (GitMensch) [inline libcob/move.c:1049]: that's seldom used - we can "inline" `n = p->times_repeated` in the condition; as `cntr_sign_neg` and `cntr_sign_pos` are only used to `break` out (if there's floating -> at least two) and these are mutually exclusive, we should change that to a single `cntr_sign`, combining the condition for -+

File: tests/testsuite.src/run_fundamental.at [hunk 2]
```diff
@@ -2510,6 +2510,102 @@ AT_CHECK([$COBCRUN_DIRECT ./prog], [0],
...
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID. prog.
+
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01 W-SRC PIC 999V99 VALUE 123.45.
+       01 W-DST PIC ZZZV.
+
+       PROCEDURE DIVISION.
+           MOVE W-SRC TO W-DST.
```
Comment (GitMensch) [inline tests/testsuite.src/run_fundamental.at:2598]: please add case of `W-SRC2 99999V9 VALUE 1234.5.` (truncation to the left)

Note: this _should_ raise

```c
/* A non-space or significant-digit was truncated during MOVE (COBOL2025) */
COB_EXCEPTION  (0308, COB_EC_DATA_TRUNCATION,
		"EC-DATA-TRUNCATION", 0)
```

which we don't have yet and also need to add for the other `MOVE` cases... this can be done with a follow-up, but should be kept in mind (at least add as a `#if 0 /* TODO: add in general` + `cob_set_exception (COB_EC_DATA_TRUNCATION);`. please - then again you may also just add it in exception.def...)

... and if/when you add that, please also add the definition for

```c
/* Attempt to use a null valued data item in a context where not permitted (COBOL2025) */
COB_EXCEPTION  (0307, COB_EC_DATA_NULL,
		"EC-DATA-NULL", 1)
```

**Review by GitMensch — CHANGES_REQUESTED** [review]
Please add left truncation in the test - I guess this goes into the same optimized function, no?

**Comment by ddeclerck** [comment]
> Please add left truncation in the test - I guess this goes into the same optimized function, no?

The optimized function deals only with same size/scale source and dest.
Added the test anyways.

**Comment by GitMensch** [comment]
I'm confused now - if both have same amount of digits and scale (which is verified in a `#if 1` block - then how does the right-side truncation occurs?

**Review by GitMensch — APPROVED** [review]
Good to commit as soon as a new Changelog entry about

`move.c (optimized_move_display_to_edited): minor refactoring`
`system.def, move.c: added definition for COPBOL2025 COB_EC_DATA_NULL + COB_EC_DATA_TRUNCATION (currently not used)

is added - but I do still find it puzzling, that the "size must match" and we seemingly still get an overflow ?!?

**Comment by GitMensch** [comment]
so... shouldn't the `#if 1` left part be triggered in the forgotten case?

In any case please adjust the test ensuring both parts are included there.

**Comment by ddeclerck** [comment]
> so... shouldn't the `#if 1` left part be triggered in the forgotten case?
> 
> In any case please adjust the test ensuring both parts are included there.

Ah, I understand what's happening. I got confused.

The `optimized_move_display_to_edited` function is indeed meant to be called on fields with same size/scale, and this seems to always be the case.

The right-truncation error message was triggered because we hit a corner case: the target PICTURE was ending with V, but the check for "desination field is full" was made *before* processing this V. Since V does not add any character to the output field, it is in fact okay if the destination is full when processing V. I'll adjust the PR.

**Comment by GitMensch** [comment]
Sounds like a good plan, as well as wrapping that check into `#ifndef NDEBUG` and adding that edge-case to the test (please do so for both the optimized and the "old" function, resulting into 4 test cases in total).

**Comment by ddeclerck** [comment]
> and adding that edge-case to the test (please do so for both the optimized and the "old" function, resulting into 4 test cases in total).

In fact, since the optimized version is only ever called when source and destination size and scale match, the only relevant test is when we have a PICTURE ending with V.
When the size and scale do not match, the move is indirect ; it is made of a call to `cob_move_display_to_display` (that does the truncations), and a call to `optimized_move_display_to_edited`.

**Comment by chuck-haatvedt** [comment]
David,

Are you doing a commit to the 3.x SVN repository for the bug 1008 fix ?

**Comment by ddeclerck** [comment]
Hi @chuck-haatvedt,
Yes, that's the plan, as soon as @GitMensch confirms this PR is OK.

**Review by GitMensch — APPROVED** [review]
That's fine with me - you likely want to replace the `#if 1` in line 1025/1035 with the same `#ifndef NDEBUG` - but that is enough on commit upstream.
